File: python/src/plotomics/hic.py
"""Hi-C contact matrix widget."""

from __future__ import annotations

from typing import Any

import numpy as np

from ._base import STATIC, PlotomicsWidget, _column, pack_columns


class HiC(PlotomicsWidget):
    """GPU-accelerated Hi-C chromatin contact map.

    The contact matrix is uploaded once as a single-channel float texture and
    drawn as one WebGL quad; the colormap and the log/linear transform run in
    the fragment shader, so pan/zoom stays smooth on very large matrices. A
    precomputed level-of-detail pyramid keeps interaction fast when zoomed out.
    Genomic coordinate ticks and the colorbar are crisp vector overlays. No
    tile server is required.

    Parameters
    ----------
    matrix:
        Either a 2-D square NumPy array of contact counts (dense), or a
        COO-like triplet of ``(i, j, v)`` for the sparse form. A triplet may be
        given as a tuple/list ``(i, j, v)``, a mapping with keys ``i``/``j``/``v``
        (e.g. a DataFrame), or via the ``i``, ``j``, ``v`` keyword arguments.
    n:
        Number of bins per axis. Required for the sparse form when it cannot be
        inferred from ``max(i, j) + 1``; ignored for a dense matrix.
    bin_size:
        Genomic bin size in base pairs, used to label axes in bp/kb/Mb. ``None``
        labels axes by bin index.
    chrom:
        Optional chromosome name shown as the axis title.
    colormap:
        Sequential colormap for intensity (currently ``"viridis"``).
    transform:
        Intensity transform, ``"log"`` (default) or ``"linear"``.
    vmax:
        Upper clip of the intensity scale; ``None`` auto-picks a high percentile.
    vmax_percentile:
        Percentile (0-1) used for the auto ``vmax`` when ``vmax`` is ``None``.
    vmin:
        Lower clip of the intensity scale.
    symmetric:
        Mirror sparse ``i``/``j``/``v`` entries across the diagonal.
    label:
        Axis title (overrides ``chrom`` when set).
    theme:
        Optional theme overrides forwarded to the JS renderer.
    height:
        Initial widget height in CSS pixels.

    Examples
    --------
    >>> import numpy as np
    >>> n = 256
    >>> d = np.abs(np.subtract.outer(np.arange(n), np.arange(n)))
    >>> m = 1000 / (d + 1) ** 1.2 + np.random.rand(n, n)
    >>> m = (m + m.T) / 2  # symmetrize
    >>> HiC(m, bin_size=10_000, chrom="chr1")  # doctest: +SKIP

    >>> # sparse COO form
    >>> HiC((np.array([0, 1]), np.array([1, 2]), np.array([5.0, 3.0])), n=3)  # doctest: +SKIP
    """

    _esm = STATIC / "hic.js"

    def __init__(
        self,
        matrix: Any = None,
        *,
        i: Any = None,
        j: Any = None,
        v: Any = None,
        n: int | None = None,
        bin_size: float | None = None,
        chrom: str | None = None,
        colormap: str = "viridis",
        transform: str = "log",
        vmax: float | None = None,
        vmax_percentile: float | None = None,
        vmin: float = 0.0,
        symmetric: bool = True,
        label: str | None = None,
        theme: dict | None = None,
        height: int = 480,
        **kwargs: Any,
    ) -> None:
        if transform not in ("log", "linear"):
            raise ValueError("`transform` must be 'log' or 'linear'.")

        columns: dict[str, Any] = {}

        # Resolve the sparse triplet from any of the accepted shapes.
        if i is None and j is None and v is None and matrix is not None:
            i, j, v = _extract_triplet(matrix)

        if i is not None and j is not None and v is not None:
            i_arr = _numeric(i, "i", np.int32)
            j_arr = _numeric(j, "j", np.int32)
            v_arr = _numeric(v, "v", np.float32)
            # Empty triplet: fail clearly instead of building an n=0 matrix.
            if i_arr.size == 0:
                raise ValueError(
                    "sparse `i`/`j`/`v` triplet is empty; provide at least one "
                    "contact."
                )
            if not (i_arr.size == j_arr.size == v_arr.size):
                raise ValueError(
                    "`i`, `j` and `v` must have the same length; "
                    f"got i={i_arr.size}, j={j_arr.size}, v={v_arr.size}"
                )
            if n is None:
                n = int(max(int(i_arr.max()), int(j_arr.max())) + 1)
            n = int(n)
            if n <= 0:
                raise ValueError("`n` must be a positive number of bins.")
            if int(i_arr.min()) < 0 or int(j_arr.min()) < 0 or \
                    int(i_arr.max()) >= n or int(j_arr.max()) >= n:
                raise ValueError(
                    f"sparse indices `i`/`j` must be in [0, {n}); got "
                    f"i in [{int(i_arr.min())}, {int(i_arr.max())}], "
                    f"j in [{int(j_arr.min())}, {int(j_arr.max())}]"
                )
            columns = {"i": i_arr, "j": j_arr, "v": v_arr}
        elif matrix is not None:
            arr = _numeric(matrix, "matrix", np.float32)
            if arr.ndim != 2 or arr.shape[0] != arr.shape[1]:
                raise ValueError("`matrix` must be a square 2-D array.")
            n = int(arr.shape[0])
            if n == 0:
                raise ValueError("`matrix` must contain at least one row and one column.")
            # Row-major (C order) flatten to match the JS `values` contract.
            columns = {"values": np.ascontiguousarray(arr).reshape(-1)}
        else:
            raise ValueError(
                "Provide a dense `matrix` or a sparse triplet (i, j, v)."
            )

        buffer, schema = pack_columns(columns)

        meta: dict[str, Any] = {"n": int(n)}
        if bin_size is not None:
            meta["binSize"] = float(bin_size)
        if chrom is not None:
            meta["chrom"] = str(chrom)

        options: dict[str, Any] = {
            "colormap": colormap,
            "transform": transform,
            "vmin": float(vmin),
            "symmetric": bool(symmetric),
        }
        # `vmax=None` means auto; only send it when the user fixed it.
        if vmax is not None:
            options["vmax"] = float(vmax)
        if vmax_percentile is not None:
            options["vmaxPercentile"] = float(vmax_percentile)
        if label is not None:
            options["label"] = str(label)
        if theme is not None:
            options["theme"] = theme

        super().__init__(
            buffer=buffer,
            schema=schema,
            data={"columns": {}, "meta": meta},
            options=options,
            _height=height,
            **kwargs,
        )


def _numeric(values: Any, name: str, dtype: Any) -> np.ndarray:
    """Coerce to a numeric array, with a clear column-named error on bad data."""
    try:
        return np.asarray(values, dtype=dtype)
    except (ValueError, TypeError):
        raise ValueError(f"`{name}` must be numeric.") from None


def _extract_triplet(matrix: Any) -> tuple[Any, Any, Any] | tuple[None, None, None]:
    """Pull (i, j, v) out of a mapping/DataFrame or a 3-tuple; else all None."""
    # tuple/list of three sequences
    if isinstance(matrix, (tuple, list)) and len(matrix) == 3:
        return matrix[0], matrix[1], matrix[2]
    # A dense numpy matrix has no named columns; it is handled by the caller.
    if isinstance(matrix, np.ndarray):
        return None, None, None
    # mapping / DataFrame with i/j/v columns
    i = _column(matrix, "i")
    j = _column(matrix, "j")
    v = _column(matrix, "v")
    if i is not None and j is not None and v is not None:
        return i, j, v
    return None, None, None

File: pkg-r/R/hic.R
#' Hi-C contact matrix
#'
#' A GPU-accelerated Hi-C chromatin contact map. The matrix is uploaded once as
#' a single-channel float texture and drawn as one WebGL quad (via `regl`), with
#' the colormap and log/linear transform applied in the fragment shader, so
#' pan/zoom stays smooth on very large matrices. A precomputed level-of-detail
#' pyramid keeps interaction fast when zoomed out. Axes (genomic coordinate
#' ticks) and the colorbar are drawn as crisp vector overlays. No tile server is
#' required.
#'
#' @param mat Either a square numeric matrix of contact counts, or a data frame
#'   / list giving a sparse COO triplet with integer columns `i`, `j` and a
#'   numeric `v`. When a triplet is supplied, `n` must be given (or inferable
#'   from `max(i, j) + 1`).
#' @param n Number of bins per axis. Required for the sparse (i/j/v) form;
#'   ignored for a dense matrix (taken from `nrow(mat)`).
#' @param bin_size Genomic bin size in base pairs; used to label axes in
#'   bp/kb/Mb. `NULL` labels axes by bin index.
#' @param chrom Optional chromosome name shown as the axis title.
#' @param colormap Sequential colormap for intensity: `"viridis"` or `"rdbu"`.
#' @param transform Intensity transform, `"log"` (default) or `"linear"`.
#' @param vmax Upper clip of the intensity scale; `NULL` auto-picks a high
#'   percentile.
#' @param vmax_percentile Percentile in `(0, 1]` used to auto-pick `vmax` when
#'   `vmax` is `NULL`. `NULL` uses the component default.
#' @param vmin Lower clip of the intensity scale.
#' @param symmetric Mirror sparse `i`/`j`/`v` entries across the diagonal.
#' @param label Axis title (overrides `chrom` when set).
#' @param theme Optional named list of theme overrides (colors, fonts, ...)
#'   merged over the component defaults in the browser. `NULL` uses the default
#'   theme.
#' @param width,height Widget dimensions (any valid CSS size).
#' @param element_id Optional explicit DOM id.
#' @return An `htmlwidget` object.
#' @examples
#' set.seed(1)
#' n <- 200
#' # distance-decay background contact matrix
#' d <- abs(outer(seq_len(n), seq_len(n), `-`))
#' m <- 1000 / (d + 1)^1.2 + matrix(runif(n * n), n, n)
#' m <- (m + t(m)) / 2 # symmetrize
#' hic(m, bin_size = 10000, chrom = "chr1")
#' @export
hic <- function(mat,
                n = NULL,
                bin_size = NULL,
                chrom = NULL,
                colormap = c("viridis", "rdbu"),
                transform = c("log", "linear"),
                vmax = NULL,
                vmax_percentile = NULL,
                vmin = 0,
                symmetric = TRUE,
                label = NULL,
                theme = NULL,
                width = NULL,
                height = NULL,
                element_id = NULL) {
  colormap <- match.arg(colormap)
  transform <- match.arg(transform)

  columns <- list()
  if (is.matrix(mat) || (is.array(mat) && length(dim(mat)) == 2)) {
    if (nrow(mat) != ncol(mat)) {
      stop("`mat` must be a square matrix.", call. = FALSE)
    }
    n <- nrow(mat)
    bv_require_nonempty(n, "mat")
    # Row-major (C order) flatten to match the JS `values` contract.
    columns$values <- bv_require_numeric(as.vector(t(mat)), "mat")
    bv_check_finite(columns$values, "mat")
  } else if (is.data.frame(mat) || is.list(mat)) {
    i <- mat[["i"]]
    j <- mat[["j"]]
    v <- mat[["v"]]
    if (is.null(i) || is.null(j) || is.null(v)) {
      stop(
        "Sparse `mat` must provide `i`, `j` and `v` columns.",
        call. = FALSE
      )
    }
    i <- bv_require_numeric(i, "i")
    j <- bv_require_numeric(j, "j")
    v <- bv_require_numeric(v, "v")
    bv_require_len(list(i = i, j = j, v = v))
    # Guard the empty-triplet trap: `max(integer(0))` is -Inf; require rows
    # before inferring `n` from the indices.
    bv_require_nonempty(length(i), "i/j/v")
    bv_check_finite(v, "v")
    i <- as.integer(i)
    j <- as.integer(j)
    if (is.null(n)) {
      n <- max(i, j) + 1L
    }
    n <- as.integer(n)
    if (any(i < 0L) || any(i >= n) || any(j < 0L) || any(j >= n)) {
      stop(sprintf(
        "hi-c `i`/`j` indices must be in [0, n) = [0, %d).", n
      ), call. = FALSE)
    }
    columns$i <- i
    columns$j <- j
    columns$v <- as.numeric(v)
  } else {
    stop(
      "`mat` must be a square matrix or a list/data frame with i/j/v.",
      call. = FALSE
    )
  }

  meta <- list(n = as.integer(n))
  if (!is.null(bin_size)) meta$binSize <- as.numeric(bin_size)
  if (!is.null(chrom)) meta$chrom <- as.character(chrom)

  options <- list(
    colormap = colormap,
    transform = transform,
    vmin = vmin,
    symmetric = isTRUE(symmetric)
  )
  # `vmax = NULL` means auto; only send it when the user fixed it.
  if (!is.null(vmax)) options$vmax <- vmax
  if (!is.null(vmax_percentile)) options$vmaxPercentile <- vmax_percentile
  if (!is.null(label)) options$label <- as.character(label)
  if (!is.null(theme)) options$theme <- theme

  plotomics_widget(
    "hic", columns,
    meta = meta,
    options = options,
    width = width, height = height, element_id = element_id
  )
}

#' Shiny bindings for hic
#'
#' Output and render functions for using [hic()] within Shiny applications and
#' interactive R Markdown / Quarto documents.
#'
#' @param output_id Output variable to read from.
#' @param width,height Element size, passed to
#'   [htmlwidgets::shinyWidgetOutput()].
#' @param expr An expression that generates a [hic()] widget.
#' @param env The environment in which to evaluate `expr`.
#' @param quoted Is `expr` already quoted? Defaults to `FALSE`.
#' @return `hicOutput()` returns a Shiny output UI element; `renderHic()`
#'   returns a Shiny render function.
#' @name hic-shiny
#' @export
hicOutput <- function(output_id, width = "100%", height = "480px") {
  htmlwidgets::shinyWidgetOutput(output_id, "hic", width, height,
    package = "plotomics"
  )
}

#' @rdname hic-shiny
#' @export
renderHic <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) {
    expr <- substitute(expr)
  }
  htmlwidgets::shinyRenderWidget(expr, hicOutput, env, quoted = TRUE)
}
